spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCdcSparkAction.java

```diff
@@ -0,0 +1,267 @@
+          newFileScanTask = ((BaseFileScanTask) fileScanTask).cloneWithoutDeletes();
+        }
+        appendedFiles.add(newFileScanTask);
+      }
+    });
+
+    return appendedFiles;
+  }
+
+  private Dataset<Row> readMetadataDeletedFiles(long snapshotId, int commitOrder) {
+    List<FileScanTask> appendedFileTasks = planMetadataDeletedFiles(snapshotId);
+    if (appendedFileTasks.isEmpty()) {
+      return null;
+    }
+
+    String groupID = UUID.randomUUID().toString();
+    FileScanTaskSetManager manager = FileScanTaskSetManager.get();
+    manager.stageTasks(table, groupID, appendedFileTasks);
+    Dataset<Row> scanDF = spark().read().format("iceberg")
+        .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
+        .load(table.name());
+
+    return withCdcColumns(scanDF, snapshotId, "D", commitOrder);
+  }
+
+  private List<FileScanTask> planMetadataDeletedFiles(long snapshotId) {
+    Snapshot snapshot = table.snapshot(snapshotId);
+    ManifestGroup manifestGroup = new ManifestGroup(table.io(), snapshot.dataManifests(), snapshot.deleteManifests())
+        .filterData(filter)
+        .ignoreAdded()
+        .specsById(((HasTableOperations) table).operations().current().specsById());
+
+    return ImmutableList.copyOf(manifestGroup.planFiles());
+  }
+
+  private Dataset<Row> readRowLevelDeletes(long snapshotId, int commitOrder) {
+    List<FileScanTask> fileScanTasks = planRowLevelDeleteFiles(snapshotId);
+    if (fileScanTasks.isEmpty()) {
+      return null;
+    }
+
+    String groupID = UUID.randomUUID().toString();
+    FileScanTaskSetManager manager = FileScanTaskSetManager.get();
+    manager.stageTasks(table, groupID, fileScanTasks);
+    Dataset<Row> scanDF = spark().read().format("iceberg")
+        .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
+        .load(table.name())
+        .filter(functions.column(MetadataColumns.IS_DELETED.name()).equalTo(true));
+
+    return withCdcColumns(scanDF, snapshotId, "D", commitOrder);
+  }
+
+  private List<FileScanTask> planRowLevelDeleteFiles(long snapshotId) {
+    Snapshot snapshot = table.snapshot(snapshotId);
+    List<ManifestFile> manifestFiles = snapshot.deleteManifests().stream()
+        .filter(manifestFile -> manifestFile.snapshotId().equals(snapshotId))
+        .collect(Collectors.toList());
+
+    // todo create a partition filter to filter out unrelated partitions
+    ManifestGroup manifestGroup = new ManifestGroup(table.io(), snapshot.dataManifests(), manifestFiles)
+        .filterData(filter)
+        .onlyWithDeletes()
+        .specsById(((HasTableOperations) table).operations().current().specsById());
+
+    return ImmutableList.copyOf(manifestGroup.planFiles());
+  }
+
+  private Dataset<Row> withCdcColumns(Dataset<Row> df, long snapshotId, String cdcType, int commitOrder) {
+    return df.withColumn(RECORD_TYPE, lit(cdcType))
+        .withColumn(COMMIT_SNAPSHOT_ID, lit(snapshotId))
```

Comment:
what are the intended usage of those 3 commit related metadata columns?

Reply:
RECORD_TYPE indicates the change types.(I, D, -U, +U)
COMMIT_SNAPSHOT_ID indicates which snapshot makes the change record.
COMMIT_TIMESTAMP indicates when the change happend.
COMMIT_ORDER indicates change order in case of multiple snapshots.